lighthouse-core/gather/gather-runner.js

```diff
@@ -108,14 +110,21 @@ class GatherRunner {
    * Navigates to about:blank and calls beforePass() on gatherers before tracing
    * has started and before navigation to the target page.
    * @param {!Object} options
+   * @param {!Object<!Array<!Promise<*>>} gathererResults
    * @return {!Promise}
    */
-  static beforePass(options) {
+  static beforePass(options, gathererResults) {
     const pass = GatherRunner.loadBlank(options.driver);
 
     return options.config.gatherers.reduce((chain, gatherer) => {
       return chain.then(_ => {
-        return gatherer.beforePass(options);
+        const artifactPromise = Promise.resolve().then(_ => gatherer.beforePass(options));
+        gathererResults[gatherer.name] = [artifactPromise];
+        return artifactPromise.catch(err => {
```

Comment:
part of the problem is that we're static here, so it's really something like `return GatherRunner.recoverOrThrow(gathererResults, gatherer.name, artifactPromise)`. I tried that, it didn't feel like a big improvement, but as you point out it does bring everything into one spot so future complications will have an easier time. What do you think?